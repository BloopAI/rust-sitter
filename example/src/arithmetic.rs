#[rust_sitter::grammar("arithmetic")]
pub mod grammar {
    #[rust_sitter::language]
    #[derive(Debug)]
    pub enum Expression {
        Number(#[rust_sitter::leaf(pattern = r"\d+", transform = |v| v.parse().unwrap())] i32),
        #[rust_sitter::prec_left(1)]
        Sub(
            Box<Expression>,
            #[rust_sitter::leaf(text = "-")] (),
            Box<Expression>,
        ),
        #[rust_sitter::prec_left(2)]
        Mul(
            Box<Expression>,
            #[rust_sitter::leaf(text = "*")] (),
            Box<Expression>,
        ),
    }

    #[rust_sitter::extra]
    struct Whitespace {
        #[rust_sitter::leaf(pattern = r"\s")]
        _whitespace: (),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn arithmetic_grammar() {
        // successful parses
        insta::assert_debug_snapshot!(grammar::parse("1"));
        insta::assert_debug_snapshot!(grammar::parse("1 - 2"));
        insta::assert_debug_snapshot!(grammar::parse("1 - 2 - 3"));
        insta::assert_debug_snapshot!(grammar::parse("1 - 2 * 3"));
        insta::assert_debug_snapshot!(grammar::parse("1 * 2 * 3"));
        insta::assert_debug_snapshot!(grammar::parse("1 * 2 - 3"));

        // failed parses
        insta::assert_debug_snapshot!(grammar::parse("1 + 2"));
        insta::assert_debug_snapshot!(grammar::parse("1 - 2 -"));
        insta::assert_debug_snapshot!(grammar::parse("a1"));
        insta::assert_debug_snapshot!(grammar::parse("1a"));
    }
}
