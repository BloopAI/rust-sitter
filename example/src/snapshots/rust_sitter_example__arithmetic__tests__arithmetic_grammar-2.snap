---
source: example/src/arithmetic.rs
expression: "grammar::parse(\"1 - 2\")"
---
Ok(
    Sub(
        Number(
            1,
        ),
        (),
        Number(
            2,
        ),
    ),
)
